Flexa-Agents/Fallback-State/src/data.fetcher.ts

```diff
@@ -0,0 +1,31 @@
+import { Contract, providers, BigNumber } from "ethers";
+import { flexaInterface } from "./abi";
+
+export default class DataFetcher {
+  readonly flexaAddress: string;
+  readonly provider: providers.Provider;
+  private flexaContract: Contract;
+
+  constructor(flexaAddress: string, provider: providers.Provider) {
+    this.flexaAddress = flexaAddress;
+    this.provider = provider;
+    this.flexaContract = new Contract(flexaAddress, flexaInterface, provider);
+  }
+
+  public async getFallbackSetDate(blockNumber: number): Promise<BigNumber> {
+    const fallbackSetDate: BigNumber = await this.flexaContract.fallbackSetDate({ blockTag: blockNumber });
+    return BigNumber.from(fallbackSetDate);
+  }
+
+  public async getFallbackWithdrawalDelaySeconds(blockNumber: number): Promise<BigNumber> {
+    const fallbackWithdrawalDelaySeconds: BigNumber = await this.flexaContract.fallbackWithdrawalDelaySeconds({
+      blockTag: blockNumber
+    });
+    return BigNumber.from(fallbackWithdrawalDelaySeconds);
+  }
+
+  public async getPreviousBlockTimestamp(blockNumber: number): Promise<number> {
+    const previousBlockTimestamp = (await this.provider.getBlock(blockNumber - 1)).timestamp;
+    return previousBlockTimestamp;
+  }

```

Comment:
I think you can change this function to `getBlockTimestamp` and pass `txEvent.blocknumber - 1` in the agent.
The thing is that there is no concept of previous block timestamp, there is only blocks timestamp. The current implementation of the Fetcher contains part agent logic due to this function.